core/lib/stages.ts

```diff
@@ -70,11 +34,47 @@ export function getStageWorkflows<T extends CommunityStage>(stages: T[]): Array<
 		}
 		return !stage.moveConstraintSources.every((constraint) => stagesById[constraint.id]);
 	});
-	// for each stage, create a list of stages that can be reached from it
-	const stageWorkflows = stageRoots.map((stage) => {
-		return createStageList(stage, stagesById);
-	});
-	return stageWorkflows as T[][];
+
+	const orderedStages = new Set<T>();
+	const stagesQueue: T[] = stageRoots;
+	const startTime = Date.now();
+	try {
+		// Breadth-first traversal of the graph(s)
+		while (stagesQueue.length > 0) {
+			if (Date.now() - startTime > 500) {
+				throw new Error("Stage sorting timed out");
+			}
```

Comment:
I think this is the only reliable way to make sure this loop stops after a timeout. I'd rather trust some other kind of global timeout but AFAIK that's not really possible in node. I'd love to know if that's wrong and there's some other technique we can use to set an enforceable global timeout

Reply:
nah i think this is fine! i was initially thinking "couldn't you use a settimeout for this?", but that will never resolve before the while loop is done, something something microtask queue something something event loop

Reply:
> something something microtask queue something something event loop

yeah this is what I meant to say. i am curious if using worker threads and/or some kind of process/fork based server would help with this in production? is that even feasible with next? i doubt this is the last time i'll accidentally add an infinite loop (or really any cpu intensive code) and i would rather not need to have timeout logic all over the app if possible!